toolchain/check/modifiers.cpp

```diff
@@ -50,28 +77,49 @@ auto ForbidModifiersOnDecl(Context& context, DeclIntroducerState& introducer,
        order_index <= static_cast<int8_t>(ModifierOrder::Last); ++order_index) {
     auto order = static_cast<ModifierOrder>(order_index);
     if (not_allowed.HasAnyOf(ModifierOrderAsSet(order))) {
-      DiagnoseNotAllowed(context, introducer.modifier_node_id(order),
-                         introducer.kind, context_string, context_loc_id);
+      DiagnoseNotAllowed(context, diagnostic_base,
+                         introducer.modifier_node_id(order), introducer.kind,
+                         context_loc_id);
       introducer.set_modifier_node_id(order, Parse::NodeId::Invalid);
     }
   }
 
   introducer.modifier_set.Remove(forbidden);
 }
 
+auto LimitModifiersOnDecl(Context& context, DeclIntroducerState& introducer,
+                          KeywordModifierSet allowed) -> void {
+  CARBON_DIAGNOSTIC(ModifierNotAllowedOnDeclaration, Error,
+                    "`{0}` not allowed on `{1}` declaration", Lex::TokenKind,
+                    Lex::TokenKind);
+  ForbidModifiersOnDecl(context, ModifierNotAllowedOnDeclaration, introducer,
+                        ~allowed);
+}
+
+auto LimitModifiersOnNotDefinition(Context& context,
+                                   DeclIntroducerState& introducer,
+                                   KeywordModifierSet allowed) -> void {
+  CARBON_DIAGNOSTIC(ModifierOnlyAllowedOnDefinition, Error,
+                    "`{0}` not allowed on `{1}` declaration, only definition",
```

Comment:
I'm a bit uncomfortable about implying that a definition isn't a kind of declaration. How would you feel about
```suggestion
                    "`{0}` not allowed on `{1}` forward declaration, only definition",
```
or maybe
```suggestion
                    "`{0}` not allowed on `{1}` non-defining declaration",
```
That said, this is entirely a drive-by (the phrasing here is unchanged from what we had before) so I'm happy for this PR to go ahead as-is and we can maybe consider this rephrasing separately.